### PR TITLE
Generate random crate names

### PR DESCRIPTION
**Pull request by scrabsha**

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -79,23 +79,31 @@ pub(crate) enum InvocationContext {
...
 impl InvocationContext {
     pub(crate) fn from_env() -> AnyResult<InvocationContext> {
         if Self::is_run_by_cargo() {
-            Ok(Self::FromCargo {
-                args: env::args().skip(1).collect(),
-                initial_version: Self::version_from_env()?,
-            })
+            let args = env::args().skip(1).collect::<Vec<_>>();
+
+            if Self::should_build_a_dependency(args.as_slice()) {
```
Comment (o0Ignition0o) [inline src/cli.rs:99]: `target_is_a_dependency` maybe ?

File: src/cli.rs [hunk 2]
```diff
@@ -79,23 +79,31 @@ pub(crate) enum InvocationContext {
...
+                Ok(InvocationContext::DepFromCargo { args })
```
Comment (o0Ignition0o) [inline src/cli.rs:100]:
```suggestion
                Ok(Self::DepFromCargo { args })
```

File: src/cli.rs [hunk 3]
```diff
@@ -79,23 +79,31 @@ pub(crate) enum InvocationContext {
...
+    /// crate.
+    GlueFromCargo {
         args: Vec<String>,
-        initial_version: Version,
+        settings: CompilerInvocationSettings,
     },
 }
 
 impl InvocationContext {
     pub(crate) fn from_env() -> AnyResult<InvocationContext> {
```
Comment (o0Ignition0o) [inline src/cli.rs:95]:
```suggestion
    pub(crate) fn from_env() -> AnyResult<Self> {
```

File: src/cli/glue_gen.rs [hunk 4]
```diff
@@ -152,6 +156,36 @@ impl GlueCrateGenerator {
...
+    fn generate_settings_file(&self, glue_path: &Path) -> AnyResult<()> {
+        let settings = self.invocation_settings();
+        Self::write_settings_file(settings, glue_path)
+    }
+
+    fn invocation_settings(&self) -> CompilerInvocationSettings {
+        CompilerInvocationSettings {
+            glue_crate_name: "glue".to_string(),
+            previous_crate_name: "previous".to_string(),
+            next_crate_name: "next".to_string(),
```
Comment (o0Ignition0o) [inline src/cli/glue_gen.rs:169]: thought we had constants somewhere ?

File: src/invocation_settings.rs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+use std::{fs, path::Path};
+
+use anyhow::{Context, Result as AnyResult};
+use semver::Version;
+use serde::{Deserialize, Serialize};
+
+/// The invocation settings.
+///
+/// When cargo-breaking is invoked by the user via CLI, cargo-breaking must
+/// generate a bunch of things. For some reasons, some must be randomely
```
Comment (o0Ignition0o) [inline src/invocation_settings.rs:10]:
```suggestion
/// generate a bunch of things. For some reasons, some must be randomly
```

"In order to make sure we won't clash with an already existing crate name when we call rustc,..."

**Comment by o0Ignition0o** [comment]
ok i think we're ready to review this, and the conflicts have (hopefully) been fixed

File: src/cli/standard_compiler.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
+use std::process::Command;
```
Comment (scrabsha) [inline src/cli/standard_compiler.rs:1]: (i'm commenting on the first line but it applies on the whole file)

I moved every compiler in `crate::compilers::*`. `StandardCompiler` can be found there.

File: src/comparator/utils.rs [hunk 7]
```diff
@@ -153,15 +154,20 @@ impl<'a> CompilationUnit<'a> {
...
             ),
-            args,
-        )?
-        .run()
+            test_compiler_settings(),
+        );
+
+        RunCompiler::new(args.as_slice(), &mut compiler)
+            .run()
+            .map_err(|_| anyhow!("Failed to compile crate"))?;
```
Comment (scrabsha) [inline src/comparator/utils.rs:168]: Ideally we should not call the compiler via `RunCompiler::new`, but rather use the `run` method defined in the `Compiler` trait. The implementation of `run` for the instrumented compiler is also responsible for handling its internal state machine, which will allow us to mark finalize as private.

This is an attempt i made to unify the behaviour of every compiler we rely on.

File: src/compiler/instrumented_compiler.rs [hunk 8]
```diff
@@ -153,3 +100,110 @@ impl Callbacks for InstrumentedCompiler {
         Compilation::Stop
     }
 }
+
```
Comment (scrabsha) [inline src/compiler/instrumented_compiler.rs:103]: Moving code here is a very cool idea. Thank you :)

File: src/compiler/instrumented_compiler.rs [hunk 9]
```diff
@@ -72,51 +50,18 @@ impl InstrumentedCompiler {
...
-        // in `StandardCompiler::run` for more.
-
-        let args = self.prepare();
-
-        RunCompiler::new(args.as_slice(), &mut self)
-            .run()
-            .map_err(|_| anyhow!("Failed to compile crate"))?;
-
-        self.finalize()
-    }
```
Comment (scrabsha) [inline src/compiler/instrumented_compiler.rs:119]: Is there a reason why we don't want this trait implemented for InstrumentedCompiler? It allowed us to cleanly handle the compiler state machine and unifies the behaviour of all our compilers :(

**Review by scrabsha — COMMENTED** [review]
Good code in this PR. I love the fact that the code you added is mostly what I would write myself.
There are some small logic errors here and there, such as the fact that we don't checkout anywhere anymore.
I also think that some older changes have been included here, as I see code I deleted recently showing up in the diff.

File: src/cli/glue_gen.rs [hunk 10]
```diff
@@ -20,138 +27,158 @@ version = "0.0.1"
...
+    temp_dir: TempDir,
 }
 
 impl GlueCrateGenerator {
-    pub(crate) fn new(package_name: String, comparison_ref: String) -> GlueCrateGenerator {
-        GlueCrateGenerator {
+    pub(crate) fn new(package_name: String, comparison_ref: String) -> AnyResult<Self> {
+        let random_suffix: &str = Word(EN).fake();
+
+        let previous_crate_name = format!("{}-{}", PREVIOUS_CRATE_NAME, random_suffix);
```
[the remark this answers is not in the document]
Comment (scrabsha) [reply src/cli/glue_gen.rs:43]: Nvm you were right

**Pull request merged**